backend/consensus/base.py

```diff
@@ -104,118 +150,113 @@ async def exec_transaction(
+                leader,
+                ExecutionMode.LEADER,
+                contract_snapshot,
+                None,
+                self.msg_handler,
             )
-            for validator in remaining_validators
-        ]
-
-        # Validators execute transaction
-        validators_results = []
-        validation_tasks = [
-            (
-                validator.exec_transaction(transaction)
-            )  # watch out! as ollama uses GPU resources and webrequest aka selenium uses RAM
-            for validator in validator_nodes
-        ]
-        validation_results = await asyncio.gather(*validation_tasks)
-
-        for i in range(len(validation_results)):
-            votes[f"{validator_nodes[i].address}"] = validation_results[i].vote.value
-        transactions_processor.update_transaction_status(
-            transaction["id"],
-            TransactionStatus.REVEALING,
-        )
 
-        if (
-            len([vote for vote in votes.values() if vote == Vote.AGREE.value])
-            < num_validators // 2
-        ):
-            raise Exception("Consensus not reached")
+            # Leader executes transaction
+            leader_receipt = await leader_node.exec_transaction(transaction)
+            votes = {leader["address"]: leader_receipt.vote.value}
+            # Update transaction status
+            transactions_processor.update_transaction_status(
+                transaction.id, TransactionStatus.COMMITTING
+            )
+
+            # Create Validators
+            validator_nodes = [
+                node_factory(
+                    validator,
+                    ExecutionMode.VALIDATOR,
+                    contract_snapshot,
+                    leader_receipt,
+                    self.msg_handler,
+                )
+                for validator in remaining_validators
+            ]
+
+            # Validators execute transaction
+            validation_tasks = [
+                (
+                    validator.exec_transaction(transaction)
+                )  # watch out! as ollama uses GPU resources and webrequest aka selenium uses RAM
+                for validator in validator_nodes
+            ]
+            validation_results = await asyncio.gather(*validation_tasks)
+
+            for i, validation_result in enumerate(validation_results):
+                votes[validator_nodes[i].address] = validation_result.vote.value
+            transactions_processor.update_transaction_status(
+                transaction.id,
+                TransactionStatus.REVEALING,
+            )
+
+            if (
+                len([vote for vote in votes.values() if vote == Vote.AGREE.value])
+                >= num_validators // 2
+            ):
+                break  # Consensus reached
+
+            print(
+                "Consensus not reached for transaction, rotating leader: ", transaction
+            )
+
+        else:  # this block is executed if the loop above is not broken
```

Comment:
the `for else` might be confusing at first, but I think it's exactly what we need here. I'm open to alternatives